### ISSUE TITLE
Zoom Region Snapshots

### ISSUE THREAD

**Issue by JBorrow** [SWIFTSIM/swiftsimio#188]
Zoom regions have a smaller (not-entirely-box-covering) grid. There will also be particles that are outside of any cell.

These will be tracked by:

- NumParticles_InCells
- NumParticles_OutsideCells

All other particles will be at the end of the array. We need to be able to read these.

- Read background particles only (i.e. outside of any cells)
- Read cell particles only (i.e. only the zoom region)
- Read sub-cells (already handled..?)
- Read cells and outside of cells (e.g. when looking at a halo right on the zoom boundary)

**Comment by MatthieuSchaller** [comment]
I will add `NumParticles_InCells_Total` and `NumParticles_InCells_ThisFile`. Same for outside. This may be helpful for distributed snapshots.

**Comment by kyleaoman** [comment]
From what I've heard of the zoom code it translates the zoom region to be in the box centre. This is a little bit annoying to compare to the parent box. There are a few ways to address this, but my suggestion after a quick think would be to (1) store some metadata in the snapshot files that contains the offset from the parent box and (2) allow a swiftsimio user to set a flag (`parent_box_coordinates=True` or something like that). This could either be once on initialisation, and then everything is assumed to happen either in the parent box frame or in the zoom box frame depending on the value, or it could be allowed to be specified on any relevant function call (e.g. anywhere a region is passed, etc).

**Comment by WillJRoper** [comment]
The current plan is to automatically shift the coordinates (and zoom region metadata) back to the parent coordinates so the outputs will always be in the same frame as the ICs (including any user-specified shift which is already stored in the snapshots).  For me, this is the simplest approach since it removes any questions about the frame. 

The bigger wrinkle is that a lot of IC generation codes will already perform the shift at which point the information is already lost (at least in my experience, I don't know if this has been addressed on more recent pipelines).

**Comment by MatthieuSchaller** [comment]
We forgot to write it here but we will undo any shift that the code adds behind the user's back. This will happen at i/o time and will put back the particles where they belong based on the ICs. 

I think that would fix the point you being up?

**Comment by kyleaoman** [comment]
Ok, if the snapshots are written in the parent box frame then there's nothing to do in swiftsimio, I think. I just wasn't sure if that was going to be straightforward with the cells involved.

**Comment by MatthieuSchaller** [comment]
@JBorrow: We currently already write the cell centre as well as the cell width. If, in the case of a zoom, I write out the centres of the zoom-region cells and their width then we have everything we need.

I can add the location of the bottom-left-front corner of the zoom region if it helps. (i.e. the (0,0,0) in the non-zoom case).

**Comment by WillJRoper** [comment]
I think we will need the bottom-left-front corner to shift during the cell index look-up? Although that could admittedly be derived from the centres and widths.

**Comment by JBorrow** [comment]
Yeah that bottom left corner will help. We won't be able to get away without any code changes here because of the background particles at the end of the array, but that's ok.

**Comment by MatthieuSchaller** [comment]
Ok. I'll add that. There might be little code changes needed to access just the zoom-in region particles though.

**Comment by WillJRoper** [comment]
@MatthieuSchaller has made some of the necessary changes (minus those needed in MPI land) on the `zoom_io` branch of SWIFT. I've tested this and generated a snapshot on cosma which can be found here:

```/cosma8/data/dp004/dc-rope1/SWIFT/test_snaps/eagle_dmo_0000.hdf5```

I've run a test on this and everything appears to be working as expected. This test can be found [here](https://github.com/WillJRoper/swift-zoom-tests/blob/main/zoom_snap_test.py) which also shows how to use the `"Origin"` attribute to extract a cell containing a position.

**Comment by MatthieuSchaller** [comment]
Some more meta-data has now been added: https://gitlab.cosma.dur.ac.uk/swift/swiftsim/-/merge_requests/2184

**Comment by kyleaoman** [comment]
The zoom code is apparently ~ready. @WillJRoper has promised me a sample snapshot soon-ish to look into what's missing to support zoom data in swiftsimio.

**Comment by JBorrow** [comment]
Very interested in getting this too for our snapshot unification project.

**Comment by WillJRoper** [comment]
Looking at this at the moment after @MatthieuSchaller tagged me in the issue about the MPI parts of this. This, of course, leads to some small headaches that I am addressing in certain configurations (mainly due to the virtual snapshot needing to concatenate zoom cell's and then background cells's particles separately to make these look-up tables work... but I digress... whoever said MPI made things simpler!). 

I will raise that shortly, but can we discuss nomenclature a bit? Having come back to things after time away, the "In" and "Out" naming is a bit vague and to me seems a bit unclear at a glance. In my opinion, it seems it would be better to have `NumParticles_ZoomCells` and `NumParticles_BkgCells` instead?

**Comment by kyleaoman** [comment]
Plenty of long names around already just spell it out `BackgroundCells`.
